moon/artifacts.py

```diff
@@ -122,44 +90,76 @@ def detect(self, left_image, right_image):
         limg = cv2.imdecode(np.fromstring(left_image, dtype=np.uint8), 1)
         rimg = cv2.imdecode(np.fromstring(right_image, dtype=np.uint8), 1)
 
-
         if self.width is None:
             self.stdout('Image resolution', limg.shape)
             self.width = limg.shape[1]
         assert self.width == limg.shape[1], (self.width, limg.shape[1])
 
 
-        ccount, w, h, x_min, x_max = count_cyan(limg[0:480,300:340])
-        if ccount > MIN_CYAN_COUNT: 
-            self.publish('artf', ['basemarker', 300, 0, 60, 480, ccount])
+        def box_area(b):
+            return b[2]*b[3]
         
         limg_rgb = cv2.cvtColor(limg, cv2.COLOR_BGR2RGB) 
         rimg_rgb = cv2.cvtColor(rimg, cv2.COLOR_BGR2RGB) 
-
-        for c in self.cascade:
-            lfound = c['classifier'].detectMultiScale(limg_rgb, minSize =(c['min_size'], c['min_size']),  maxSize =(c['max_size'], c['max_size'])) 
-            rfound = c['classifier'].detectMultiScale(rimg_rgb, minSize =(c['min_size'], c['min_size']),  maxSize =(c['max_size'], c['max_size'])) 
-
-            if len(lfound) > 0 and len(rfound) > 0: # only report if both cameras see it
-                if c['subsequent_detects'] < 3: # do not act until you have at least 3 detections in a row
-                    c['subsequent_detects'] += 1
+        hsv = cv2.cvtColor(limg, cv2.COLOR_BGR2HSV)
+        hsv_blurred = cv2.medianBlur(hsv,5) # some frames have noise, need to blur otherwise threshold doesn't work
+
+        for c in self.detectors:
+            if c['artefact_name'] not in self.detect_sequences:
+                self.detect_sequences[c['artefact_name']] = 0
+
+            if c['detector_type'] == 'colormatch':
+                lower_hue = np.array([c['hue_match'] - c['hue_max_difference'],50,50])
+                upper_hue = np.array([c['hue_match'] + c['hue_max_difference'],255,255])
+                # Threshold the HSV image to get only the matching colors
+                mask = cv2.inRange(hsv_blurred, lower_hue, upper_hue)
+                mser = cv2.MSER_create(_min_area=c['pixel_count_threshold'])
+
+                _, bboxes = mser.detectRegions(mask)
```

Comment:
it would be nice to have some test examples

Reply:
This deals with images, are you suggesting to encode a JPEG in text of some test script? Or maybe elaborate what the input and output of such test should be? One question I would have is that finding a bounding box doesn't necessarily have a single correct answer, there will be pixels on the margin which can be equally reasonably included and excluded. So when some parameters are adjusted in the code later and a few pixels fall out when they were previously in, it doesn't necessarily mean the test failed. Maybe the test is for the bbox to be near but not necessarily exact match with the expected response?

Reply:
I would save just one or two examples of images. And yes, you can use some tolerance if necessary ... it is just "placeholder" for really strange examples or bugs in the future.

Reply:
test case created for all 3 currently visually recognized objects